File: frontend/src/types/dependencies.ts
export type UsedBySeverity = "blocking" | "warning" | "info";

export interface UsedByDependent {
  severity: UsedBySeverity;
  dependent_type: string;
  dependent_id: string | null;
  dependent_label: string | null;
  location: string | null;
  message: string;
  recommendation: string | null;
  column_id?: string | null;
  column_label?: string | null;
  rule_id?: string | null;
  rule_label?: string | null;
  cell_key?: string | null;
  related_id?: string | null;
  related_type?: string | null;
}

export interface UsedByReport {
  resource_type: string;
  resource_id: string;
  resource_label: string | null;
  safe_to_delete: boolean;
  blocking_count: number;
  warning_count: number;
  info_count: number;
  dependents: UsedByDependent[];
}

File: backend/app/schemas/dependencies.py
"""Shared dependency/used-by response shapes.

These schemas power destructive-action guards. They are deliberately
generic so Reference Data, Import Builder, and Invoice Builder can all
return the same report shape without coupling their own schemas.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


UsedBySeverity = Literal["blocking", "warning", "info"]


class UsedByDependent(BaseModel):
    severity: UsedBySeverity
    dependent_type: str = Field(min_length=1, max_length=80)
    dependent_id: str | None = Field(default=None, max_length=128)
    dependent_label: str | None = Field(default=None, max_length=255)
    location: str | None = Field(default=None, max_length=255)
    message: str = Field(min_length=1, max_length=500)
    recommendation: str | None = Field(default=None, max_length=500)
    column_id: str | None = Field(default=None, max_length=128)
    column_label: str | None = Field(default=None, max_length=255)
    rule_id: str | None = Field(default=None, max_length=128)
    rule_label: str | None = Field(default=None, max_length=255)
    cell_key: str | None = Field(default=None, max_length=128)
    related_id: str | None = Field(default=None, max_length=128)
    related_type: str | None = Field(default=None, max_length=80)


class UsedByReport(BaseModel):
    resource_type: str = Field(min_length=1, max_length=80)
    resource_id: str = Field(min_length=1, max_length=128)
    resource_label: str | None = Field(default=None, max_length=255)
    safe_to_delete: bool
    blocking_count: int = 0
    warning_count: int = 0
    info_count: int = 0
    dependents: list[UsedByDependent] = Field(default_factory=list)
